File: app/src/main/java/com/bis5/fitjourney/fragments/ProgressFragment.java
package com.bis5.fitjourney.fragments;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.NavigationUI;

import com.bis5.fitjourney.databinding.FragmentProgressBinding;
import com.bis5.fitjourney.models.AppDatabase;
import com.bis5.fitjourney.models.WorkoutLog;
import com.bis5.fitjourney.viewmodels.WorkoutViewModel;
import com.bis5.fitjourney.viewmodels.WorkoutViewModelFactory;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;

import java.util.ArrayList;
import java.util.List;

public class ProgressFragment extends Fragment {

    private FragmentProgressBinding binding;
    private WorkoutViewModel workoutViewModel;
    private long workoutId; // THE LAW IS OBEYED

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentProgressBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (getArguments() != null) {
            // THE REBELLION IS CRUSHED. THE GHOST IS BANISHED.
            workoutId = ProgressFragmentArgs.fromBundle(getArguments()).getWorkoutId();
        }

        NavController navController = NavHostFragment.findNavController(this);
        NavigationUI.setupWithNavController(binding.toolbar, navController);
        binding.toolbar.setTitle("Workout Progress");

        AppDatabase database = AppDatabase.getDatabase(requireContext());
        WorkoutViewModelFactory viewModelFactory = new WorkoutViewModelFactory(database.workoutDao(), database.exerciseDao(), database.workoutLogDao(), database.setLogDao());
        workoutViewModel = new ViewModelProvider(this, viewModelFactory).get(WorkoutViewModel.class);

        setupChart();

        workoutViewModel.getWorkoutHistory(workoutId).observe(getViewLifecycleOwner(), history -> {
            if (history != null && !history.isEmpty()) {
                List<Entry> entries = new ArrayList<>();
                for (int i = 0; i < history.size(); i++) {
                    WorkoutLog workoutLog = history.get(i);
                    double volume = 0.0;
                    if (workoutLog.getDateFinished() != null) {
                        volume = (workoutLog.getDateFinished() - workoutLog.getDateStarted()) / 1000.0;
                    }
                    entries.add(new Entry((float) i, (float) volume));
                }

                LineDataSet dataSet = new LineDataSet(entries, "Total Volume Over Time");
                dataSet.setColor(Color.BLUE);
                dataSet.setValueTextColor(Color.BLACK);

                LineData lineData = new LineData(dataSet);
                binding.lineChart.setData(lineData);
                binding.lineChart.invalidate();
            }
        });
    }

    private void setupChart() {
        binding.lineChart.getDescription().setEnabled(false);
        binding.lineChart.getXAxis().setPosition(XAxis.XAxisPosition.BOTTOM);
        binding.lineChart.getAxisRight().setEnabled(false);
        binding.lineChart.setDrawGridBackground(false);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/bis5/fitjourney/fragments/ProgressFragment.kt
package com.bis5.fitjourney.fragments

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import androidx.navigation.ui.setupWithNavController
import com.bis5.fitjourney.databinding.FragmentProgressBinding
import com.bis5.fitjourney.models.AppDatabase
import com.bis5.fitjourney.viewmodels.WorkoutViewModel
import com.bis5.fitjourney.viewmodels.WorkoutViewModelFactory
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.ValueFormatter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ProgressFragment : Fragment() {

    private var _binding: FragmentProgressBinding? = null
    private val binding get() = _binding!!

    private lateinit var workoutViewModel: WorkoutViewModel
    private val args: ProgressFragmentArgs by navArgs()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentProgressBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val navController = findNavController()
        binding.toolbar.setupWithNavController(navController)
        binding.toolbar.title = "Workout Progress"

        val database = AppDatabase.getDatabase(requireContext())
        val viewModelFactory = WorkoutViewModelFactory(database.workoutDao(), database.exerciseDao(), database.workoutLogDao(), database.setLogDao())
        workoutViewModel = ViewModelProvider(this, viewModelFactory)[WorkoutViewModel::class.java]

        setupChart()

        workoutViewModel.getWorkoutHistory(args.workoutId).observe(viewLifecycleOwner) { history ->
            if (history.isNotEmpty()) {
                val entries = history.mapIndexed { index, workoutLog ->
                    // Calculate total volume for this workout log
                    // This is a simplified approach. A more robust solution would involve
                    // fetching the sets for each log and summing them up.
                    val volume = workoutLog.dateFinished?.let { (it - workoutLog.dateStarted) / 1000.0 } ?: 0.0
                    Entry(index.toFloat(), volume.toFloat())
                }

                val dataSet = LineDataSet(entries, "Total Volume Over Time")
                dataSet.color = Color.BLUE
                dataSet.valueTextColor = Color.BLACK

                val lineData = LineData(dataSet)
                binding.lineChart.data = lineData
                binding.lineChart.invalidate() // Refresh the chart
            }
        }
    }

    private fun setupChart() {
        binding.lineChart.apply {
            description.isEnabled = false
            xAxis.position = XAxis.XAxisPosition.BOTTOM
            axisRight.isEnabled = false
            setDrawGridBackground(false)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
